worlds/kh2/Rules.py

```diff
@@ -1,96 +1,1839 @@
+        final_leveling_access = [LocationName.MemorysSkyscaperMythrilCrystal, LocationName.GrimReaper2,
+                                 LocationName.Xaldin, LocationName.StormRider, LocationName.SunsetTerraceAbilityRing]
+        return any(
+                self.world.multiworld.get_location(location, self.player).can_reach(state) for location in
+                final_leveling_access
+        )
+
+    def limit_form_region_access(self, state: CollectionState) -> bool:
+        """
+        multi_form_region_access + namine sketches for leveling
+        """
+        multi_form_region_access = {
+            ItemName.CastleKey,
+            ItemName.BattlefieldsofWar,
+            ItemName.SwordoftheAncestor,
+            ItemName.BeastsClaw,
+            ItemName.BoneFist,
+            ItemName.SkillandCrossbones,
+            ItemName.Scimitar,
+            ItemName.MembershipCard,
+            ItemName.IceCream,
+            ItemName.WaytotheDawn,
+            ItemName.IdentityDisk,
+            ItemName.NamineSketches
+        }
+        return state.has_any(multi_form_region_access, self.player)
+
+    def multi_form_region_access(self, state: CollectionState) -> bool:
+        """
+        Valor, Wisdom and Master Form region access.
+        Note: This does not account for having the drive form. See drive_form_unlock
+        """
+        # todo: if boss enemy start the player with oc stone because of cerb
+        multi_form_region_access = {
+            ItemName.CastleKey,
+            ItemName.BattlefieldsofWar,
+            ItemName.SwordoftheAncestor,
+            ItemName.BeastsClaw,
+            ItemName.BoneFist,
+            ItemName.SkillandCrossbones,
+            ItemName.Scimitar,
+            ItemName.MembershipCard,
+            ItemName.IceCream,
+            ItemName.WaytotheDawn,
+            ItemName.IdentityDisk,
+        }
+        return state.has_any(multi_form_region_access, self.player)
+
+    def level_locking_unlock(self, state: CollectionState, amount):
+        return amount <= sum([state.count(item_name, self.player) for item_name in visit_locking_dict["2VisitLocking"]])
+
+    def summon_levels_unlocked(self, state: CollectionState, amount) -> bool:
+        return amount <= sum([state.count(item_name, self.player) for item_name in summons])
+
+    def set_kh2_rules(self) -> None:
+        world = self.world.multiworld
+        player = self.player
+        for region in world.get_regions(self.player):
+            if region.name in self.region_rules:
+                for entrance in region.entrances:
+                    entrance.access_rule = self.region_rules[region.name]
+
+        self.set_kh2_goal()
+
+        for slot, weapon in exclusion_table["WeaponSlots"].items():
+            add_rule(world.get_location(slot, player), lambda state: state.has(weapon, player))
+
+        #  Forbid Abilities on popups due to game limitations
+        for location in list(popups_set):
```

Comment:
all of these separate location lists you have... you should probably have them defined elsewhere (maybe module level?) then reference them in the original `for region in multiworld.get_regions(self.player)` loop. that way you're looping over each region and location a single time and adjusting the rules accordingly instead of looping over every single location multiple times which is very slow.